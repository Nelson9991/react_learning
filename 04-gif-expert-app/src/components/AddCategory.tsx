import { ChangeEvent, useState } from 'react';

interface Props {
  onNewCategory: (newCategory: string) => void;
}

export const AddCategory = ({ onNewCategory }: Props) => {
  const [inputValue, setInputValue] = useState('');

  const handleInputChange = ({ target }: ChangeEvent<HTMLInputElement>) => {
    setInputValue(target.value);
  };

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (inputValue.trim().length <= 1) return;

    onNewCategory(inputValue.trim());
    setInputValue('');
  };

  return (
    <form onSubmit={(e) => handleSubmit(e)}>
      <input
        type='text'
        placeholder='Buscar gifs'
        value={inputValue}
        onChange={handleInputChange}
      />
    </form>
  );
};
