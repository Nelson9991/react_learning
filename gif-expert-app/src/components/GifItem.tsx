import { GifResponse } from '../models/gif-response.model';

interface Props {
  gif: GifResponse;
}

export const GifItem = ({ gif }: Props) => {
  return (
    <div className='card'>
      <img src={gif.url} alt={gif.title} />
      <p>{gif.title}</p>
    </div>
  );
};

export default GifItem;
