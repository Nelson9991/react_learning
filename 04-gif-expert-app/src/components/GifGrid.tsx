import GifItem from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

interface Props {
  category: string;
}

export const GifGrid = ({ category }: Props) => {
  const { gifs, isLoading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {
        // If isLoading is true, then show the loading message
        isLoading && <p>Loading...</p>
      }

      <div className='card-grid'>
        {gifs.map((gif) => (
          <GifItem key={gif.id} gif={gif} />
        ))}
      </div>
    </>
  );
};

export default GifGrid;
