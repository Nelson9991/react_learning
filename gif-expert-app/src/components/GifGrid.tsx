import GifItem from './GifItem';
import { useFecthGifs } from '../hooks/useFecthGifs';

interface Props {
  category: string;
}

export const GifGrid = ({ category }: Props) => {
  const { gifs, isLoading } = useFecthGifs(category);

  return (
    <>
      <h3>{category}</h3>
      {
        // If isLoading is true, then show the loading message
        isLoading && <p>Loading...</p>
      }

      <div className='card-grid'>
        {gifs.map((gif) => (
          <GifItem key={gif.id} gif={gif} />
        ))}
      </div>
    </>
  );
};

export default GifGrid;
