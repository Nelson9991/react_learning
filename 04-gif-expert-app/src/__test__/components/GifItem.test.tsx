import { render, screen } from '@testing-library/react';
import { GifItem } from '../../components';
import { GifResponse } from '../../models/gif-response.model';

describe('Test in <GifItem/>', () => {
  const gif: GifResponse = {
    id: 'ABC',
    title: 'A title',
    url: 'https://localhost/any/thing.jpg'
  };

  test('snapshot must match', () => {
    const { container } = render(<GifItem gif={gif} />);
    expect(container).toMatchSnapshot();
  });

  test('should render an img with proper src and alt', () => {
    render(<GifItem gif={gif} />);
    // screen.debug(); // sirve para ver el contenido del render()
    // expect((screen.getByRole('img') as HTMLImageElement).src).toBe(gif.url);
    // expect((screen.getByRole('img') as HTMLImageElement).alt).toBe(gif.title);
    const { src, alt } = screen.getByRole('img') as HTMLImageElement;
    expect(src).toBe(gif.url);
    expect(alt).toBe(gif.title);
  });

  test('should render the gif title in component', () => {
    render(<GifItem gif={gif} />);
    expect(screen.getByText(gif.title)).toBeTruthy();
  });
});
