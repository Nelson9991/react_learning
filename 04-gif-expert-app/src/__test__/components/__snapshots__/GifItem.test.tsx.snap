// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <GifItem/> snapshot must match 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="A title"
      src="https://localhost/any/thing.jpg"
    />
    <p>
      A title
    </p>
  </div>
</div>
`;
